Add logging for async compile worker statistics

aorenste opened the pull request:
Add some on-exit logging to the async compile workers. When you use `TORCH_LOGS=async_compile` (or `all`) it will now report how many workers were enqueued & dequeued (should be the same) as well as queuing time (how long workers sat on the queue before starting to run) and maximum depth (how many workers were waiting to start.

Tested manually by running a larger internal model and then lowering the number of available workers to see the time and depth get longer.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #155820



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

aorenste committed:
Add logging for async compile worker statistics

[ghstack-poisoned]

aorenste committed:
Update on "Add logging for async compile worker statistics"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

aorenste commented on torch/_inductor/async_compile.py:
```diff
@@ -515,6 +519,89 @@ def _wait_futures(self, scope: dict[str, Any]) -> None:
+
+class TrackedProcessPoolExecutor(ProcessPoolExecutor):
+    # NOTE: Use self._shutdown_lock to control access to this data
+    _queue_stats: _QueueStats
```
oops - need to remove this

aorenste committed:
Update on "Add logging for async compile worker statistics"

Add some on-exit logging to the async compile workers. When you use `TORCH_LOGS=async_compile` (or `all`) it will now report how many workers were enqueued & dequeued (should be the same) as well as queuing time (how long workers sat on the queue before starting to run) and maximum depth (how many workers were waiting to start.

Tested manually by running a larger internal model and then lowering the number of available workers to see the time and depth get longer.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

masnesral commented on torch/_inductor/async_compile.py:
```diff
@@ -515,6 +519,87 @@ def _wait_futures(self, scope: dict[str, Any]) -> None:
+        super().__init__(*args, **kwargs)
+
+    def _record_dequeue(self, f: concurrent.futures.Future[Any]) -> None:
+        now = time()
```
For some of the scuba logging work around timing, I think there were concerns about time.time() not being accurate when measuring at the level of _us, so we standardized on time.time_ns() (and then divided to get _us). Not sure if it's a concern here, but FYI we're using time_ns() now in most places.

aorenste replied:
I hope we're not using this for that level of timing. My intent was #1 to figure out how deep the queue got (at the worst time do we have 4 things in the queue or 400?) and then to get a rough idea of timing (things sitting in the queue for 0.1s vs 0.09s doesn't matter - sitting in the queue for 50s vs 3s is more important to note).

masnesral commented on torch/_inductor/compile_worker/subproc_pool.py:
```diff
@@ -269,7 +269,9 @@ def __init__(
 
     def _new_pool(self, nprocs: int, warm: bool) -> ProcessPoolExecutor:
-        pool = ProcessPoolExecutor(
+        from torch._inductor.async_compile import TrackedProcessPoolExecutor
```
nit: Maybe the TrackedProcessPoolExecutor impl belongs in this file anyway? I dunno.

aorenste replied:
Yeah - that's probably a good idea. I'll move it.

aorenste replied:
actually - I'll move it to it's own file under compile_worker

masnesral commented on torch/_inductor/async_compile.py:
```diff
@@ -515,6 +519,87 @@ def _wait_futures(self, scope: dict[str, Any]) -> None:
+_queue_stats_lock = threading.Lock()
+
+
+class TrackedProcessPoolExecutor(ProcessPoolExecutor):
```
I'm working on a PR now that would shut down the pool when compilation ends and start a new one on the next compile. I don't think that's a problem for anything here, but just double checking.

aorenste committed:
Update on "Add logging for async compile worker statistics"

Add some on-exit logging to the async compile workers. When you use `TORCH_LOGS=async_compile` (or `all`) it will now report how many workers were enqueued & dequeued (should be the same) as well as queuing time (how long workers sat on the queue before starting to run) and maximum depth (how many workers were waiting to start.

Tested manually by running a larger internal model and then lowering the number of available workers to see the time and depth get longer.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

aorenste commented on torch/_logging/_registrations.py:
```diff
@@ -13,6 +13,7 @@
     "torch.nn.parallel.distributed",
 ]
 
+register_log("async_compile", ["torch._inductor.async_compile"])
```
Nice catch...

aorenste committed:
Update on "Add logging for async compile worker statistics"

Add some on-exit logging to the async compile workers. When you use `TORCH_LOGS=async_compile` (or `all`) it will now report how many workers were enqueued & dequeued (should be the same) as well as queuing time (how long workers sat on the queue before starting to run) and maximum depth (how many workers were waiting to start.

Tested manually by running a larger internal model and then lowering the number of available workers to see the time and depth get longer.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "Add logging for async compile worker statistics"

Add some on-exit logging to the async compile workers. When you use `TORCH_LOGS=async_compile` (or `all`) it will now report how many workers were enqueued & dequeued (should be the same) as well as queuing time (how long workers sat on the queue before starting to run) and maximum depth (how many workers were waiting to start.

Tested manually by running a larger internal model and then lowering the number of available workers to see the time and depth get longer.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

aorenste commented:
@pytorchbot merge
